src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineControl.xaml.cs

```diff
@@ -0,0 +1,330 @@
+                    }
+
+                    return;
+                }
+
+                delay.Reset();
+            }
+
+            async Task UpdateAsync()
+            {
+                var response = await DocumentSymbolsRequestAsync(textBuffer, languageServiceBroker).ConfigureAwait(false);
+                if (response is not null && response.Response is not null)
+                {
+                    await threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(CancellationToken.None);
+                    var responseBody = response.Response.ToObject<DocumentSymbol[]>();
+                    var documentSymbols = DocumentOutlineHelper.GetNestedDocumentSymbols(responseBody);
+                    var documentSymbolModels = DocumentOutlineHelper.GetDocumentSymbolModels(documentSymbols);
+                    SymbolTreeInitialized = true;
+                    SymbolsTreeItemsSource = documentSymbolModels;
+                    symbolTree.ItemsSource = documentSymbolModels;
+                }
+                else
+                {
+                    symbolTree.ItemsSource = new List<DocumentSymbolViewModel>();
+                    SymbolsTreeItemsSource = new List<DocumentSymbolViewModel>();
+                }
+            }
+        }
+
+        private async Task<ManualInvocationResponse?> DocumentSymbolsRequestAsync(ITextBuffer textBuffer, ILanguageServiceBroker2 languageServiceBroker)
+        {
+            var parameterFactory = new DocumentSymbolParams()
+            {
+                TextDocument = new TextDocumentIdentifier()
+                {
+                    Uri = new Uri(_filePath)
+                }
+            };
+
+            var serializedParams = JToken.FromObject(parameterFactory);
+            JToken ParameterFactory(ITextSnapshot _)
+            {
+                return serializedParams;
+            }
+
+            // TODO: proper workaround such that context.ClientCapabilities?.TextDocument?.DocumentSymbol?.HierarchicalDocumentSymbolSupport == true
+            return await languageServiceBroker.RequestAsync(
+                textBuffer: textBuffer,
+                method: Methods.TextDocumentDocumentSymbolName,
+                capabilitiesFilter: (JToken x) => true,
+                languageServerName: WellKnownLspServerKinds.AlwaysActiveVSLspServer.ToUserVisibleString(),
+                parameterFactory: ParameterFactory,
+                cancellationToken: CancellationToken.None).ConfigureAwait(false);
+        }
+
+        private static string? GetFilePath(IWpfTextView textView)
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+            if (textView.TextBuffer.Properties.TryGetProperty(typeof(IVsTextBuffer), out IVsTextBuffer bufferAdapter) &&
+                bufferAdapter is IPersistFileFormat persistFileFormat &&
+                ErrorHandler.Succeeded(persistFileFormat.GetCurFile(out var filePath, out _)))
+            {
+                return filePath;
+            }
```

Comment:
what do we need the file-path for?

Reply:
It's needed to make the LSP document symbol request: I need to pass it into `languageServiceBroker.RequestAsync`